extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/client/VertxVaultClient.java

```diff
@@ -377,143 +386,147 @@ public void deletePolicy(String token, String name) {
 
     protected int head(String path) {
-        final Request request = builder(path).head().build();
+        final HttpRequest<Buffer> request = builder(path).method(HttpMethod.HEAD);
         return exec(request);
     }
 
     protected int head(String path, Map<String, String> queryParams) {
-        final Request request = builder(path, queryParams).head().build();
+        final HttpRequest<Buffer> request = builder(path, queryParams).method(HttpMethod.HEAD);
         return exec(request);
     }
 
-    private <T> T exec(Request request, Class<T> resultClass) {
-        return exec(request, resultClass, 200);
+    private <T> T exec(HttpRequest<Buffer> request, Class<T> resultClass) {
+        return exec(request, null, resultClass, 200);
     }
 
-    private <T> T exec(Request request, int expectedCode) {
-        return exec(request, null, expectedCode);
+    private <T> T exec(HttpRequest<Buffer> request, int expectedCode) {
+        return exec(request, null, null, expectedCode);
     }
 
-    private <T> T exec(Request request, Class<T> resultClass, int expectedCode) {
-        try (Response response = client.newCall(request).execute()) {
-            if (response.code() != expectedCode) {
+    private <T> T exec(HttpRequest<Buffer> request, Object body, Class<T> resultClass) {
+        return exec(request, body, resultClass, 200);
+    }
+
+    private <T> T exec(HttpRequest<Buffer> request, Object body, Class<T> resultClass, int expectedCode) {
+        try {
+            HttpResponse<Buffer> response = body == null
+                    ? request.sendAndAwait()
+                    : request.sendBufferAndAwait(Buffer.buffer(requestBody(body)));
+            if (response.statusCode() != expectedCode) {
                 throwVaultException(response);
             }
-            String jsonBody = response.body().string();
-            return resultClass == null ? null : mapper.readValue(jsonBody, resultClass);
-        } catch (IOException e) {
+            Buffer responseBuffer = response.body();
+            if (responseBuffer != null) {
+                return resultClass == null ? null : mapper.readValue(responseBuffer.toString(), resultClass);
+            } else {
+                return null;
+            }
+        } catch (JsonProcessingException e) {
             throw new VaultException(e);
         }
     }
 
-    private int exec(Request request) {
-        try (Response response = client.newCall(request).execute()) {
-            return response.code();
-        } catch (IOException e) {
-            throw new VaultException(e);
-        }
+    private int exec(HttpRequest<Buffer> request) {
+        return request.sendAndAwait().statusCode();
     }
 
-    private void throwVaultException(Response response) {
+    private void throwVaultException(HttpResponse<Buffer> response) {
         String body = null;
         try {
-            body = response.body().string();
+            body = response.body().toString();
```

Comment:
Is encoding / charset something to handle here?

Reply:
to be honest I am not sure. I assumed `UTF-8`
I could not find some info in the vault doc.
if I curl vault, I get something such as:
```
$ curl -v http://localhost:8200/v1/sys/health
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8200 (#0)
> GET /v1/sys/health HTTP/1.1
> Host: localhost:8200
> User-Agent: curl/7.54.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Cache-Control: no-store
< Content-Type: application/json
< Date: Fri, 22 Jan 2021 11:53:00 GMT
< Content-Length: 294
< 
{"initialized":true,"sealed":false,"standby":false,"performance_standby":false,"replication_performance_mode":"disabled","replication_dr_mode":"disabled","server_time_utc":1611316380,"version":"1.5.4","cluster_name":"vault-cluster-6dcd7526","cluster_id":"2e1e111c-7ac6-2499-358a-cd7402808803"}
* Connection #0 to host localhost left intact
```

so no encoding. should we assume `UTF-8` then?

in the [vault tests](https://github.com/quarkusio/quarkus/blob/master/test-framework/vault/src/main/java/io/quarkus/vault/test/VaultTestExtension.java#L69), I had added a `euro` to try to detect encoding issues (the secret is written using the cli, and read from the api). If we had an issue with encoding, I think this should have failed.

do you think we can do better?

Reply:
That's good enough for now. If we find out that encoding needs to be handled or configured, we can do that later.